lessons/advanced/typespec.md

```diff
@@ -0,0 +1,197 @@
+
+```elixir
+@spec sum_times(integer) :: integer
+def sum_times(a) do
+    [1, 2, 3]
+    |> Enum.map(fn el -> el * a end)
+    |> Enum.sum
+    |> round
+end
+```
+
+Finally:
+
+```shell
+$ mix dialyzer
+...
+  Proceeding with analysis... done in 0m0.95s
+done (passed successfully)
+```
+
+Using specifications with tools to static code analysis helps us to make code that is self tested and contains less bugs.  
+
+## Custom types
+
+Writing specification is nice, but sometimes our functions works with more complex data structures than simple numbers or collections. In that cases definition in `@spec` could be hard to understand and very hard to change by other developers. Sometimes functions needs to take many parameters or return complex data. Long parameters list is one of bad smells in code. In object oriented languages like Ruby or Java we could easily define classes that helps us to solve this problems. Elixir hasn't classes but because is easy to extends that we could define our types.
+  
+Out of box Elixir contains some basic types like `integer` or `pid`. You  can find full list of available types in [documentation](http://elixir-lang.org/docs/stable/elixir/typespecs.html#types-and-their-syntax).
+ 
+### Defining custom type
+  
+Let's modify our `sum_times` function and introduce some extra params:
+
+```elixir
+@spec sum_times(integer, %Examples{first: integer, last: integer}) :: integer
+def sum_times(a, params) do
+    for i <- params.first..params.last do
+        i
+    end
+       |> Enum.map(fn el -> el * a end)
+       |> Enum.sum
+       |> round
+end
+```
+
+We introduce struct in `Examples` module that contains two fields `first` and `last`. That is simpler version of struct from `Range` module. About structures we talk when discussing [modules](lessons/basics/modules/#structs). Lets imagine that we need to specification with `Examples` struct many times. It will be very annoying and could be source of bugs. Solution of this problem is `@type`.
+ 
+Elixir has three directives for types:
+
+  - `@type` – simple, public type. Internal structure of type is public. 
+  - `@typep` – type is private and could be used only in module where is defined. 
+  - `@opaque` – type is public, but internal structure is private. 
+
+Let define our type:
+
+```elixir
+defmodule Examples do
+
+    defstruct first: nil, last: nil
+
+    @type t :: %Examples{first: integer, last: integer}
+
+    @type t(first, last) :: %Examples{first: first, last: last}
+
+end
+```
+
+We defined two types. First `t` is representation of `%Examples{first: integer, last: integer}`. Second is `t(first, last)` that is an alias to `%Examples{first: first, last: last}`. What is a difference? First one represents struct `Examples` of which the two keys are `integer`. Second one represents struct which keys could has any type. That means code like this:
```

Comment:
Need to rewrite this. After few days I know what I wanted to write, but i wrote it in terrible way.